### PR TITLE
Added training code and wandb support

### PR DESCRIPTION
**Pull request by Jazhyc**
- Training and evaluation loops work
- Refactored init method of trainer to have different setup methods
- Added wandb support to keep track of runs (requires .env file with WANDB_API_KEY and WANDB_ENTITY=im2latex-replicate). Currently just measures loss and BLEU score
- Removed distributed support (I did not see Matthijs' message earlier and I wanted to keep the code simpler)

TODO
- Refactor training and evaluation loop (I kept these mainly the same as they were but they could be simplified)
- Train model
- Decide on how checkpointing should work. Each save takes 1GB of space so we should be careful.

### CODE REVIEW

File: im2latex/trainers/visionencoderdecodertrainer.py [hunk 1]
```diff
@@ -203,108 +229,97 @@ def train(self) -> 'AbstractTrainer':
...
-
                 num_evaluated_batches += 1
 
         self.avg_val_loss = np.mean(val_losses)
         self.avg_bleu = np.mean(bleu_scores)
 
-        # Do something with these losses and scores
-        return self
+        # Return metrics
+        return self.avg_val_loss, self.avg_bleu
```
Comment (matthjs) [inline im2latex/trainers/visionencoderdecodertrainer.py:322]: Need to adjust function signature since it returns a tuple now

File: im2latex/trainers/visionencoderdecodertrainer.py [hunk 2]
```diff
@@ -23,107 +23,142 @@
...
+        
+        self.model.config.max_length = self.cfg.decoding.max_length
+        self.model.config.early_stopping = self.cfg.decoding.early_stopping
+        self.model.config.num_beams = self.cfg.decoding.num_beams
+        
         self.model.decoder.resize_token_embeddings(len(self.tokenizer))
 
-        # TODO: Make configurable or/and probably want to change this.
         self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        print(f"Using device: {self.device}")
```
Comment (matthjs) [inline im2latex/trainers/visionencoderdecodertrainer.py]: Maybe use the logger instead?

**Review by matthjs — APPROVED** [review]
Looks good so far, regarding the distributed support we can also leave it out for now. We can do further refactor ones all the functionality is in place.
